cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowSpecification.java

```diff
@@ -32,21 +34,28 @@
   private final String className;
   private final String name;
   private final String description;
-  private final Map<String, WorkflowActionSpecification> customActionMap;
-  private final List<ScheduleProgramInfo> actions;
   private final Map<String, String> properties;
 
+  private List<WorkflowNode> nodes = Lists.newArrayList();
+  private Map<String, WorkflowForkSpecification> forks = Maps.newHashMap();
+  private Map<String, ScheduleProgramInfo> actions = Maps.newHashMap();
+  private Map<String, WorkflowActionSpecification> customActionMap = Maps.newHashMap();
+
   public WorkflowSpecification(String className, String name, String description,
-                                      Map<String, String> properties,
-                                      List<ScheduleProgramInfo> actions,
+                                      Map<String, String> properties, List<WorkflowNode> nodes,
+                                      Map<String, WorkflowForkSpecification> forks,
+                                      Map<String, ScheduleProgramInfo> actions,
                                       Map<String, WorkflowActionSpecification> customActionMap) {
     this.className = className;
     this.name = name;
     this.description = description;
     this.properties = properties == null ? Collections.<String, String>emptyMap() :
                                            Collections.unmodifiableMap(new HashMap<String, String>(properties));
-    this.actions = Collections.unmodifiableList(new ArrayList<ScheduleProgramInfo>(actions));
```

Comment:
Having the class organization based on limitation serialization seems not a good practice.